x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -145,15 +145,39 @@ func (k Keeper) FinalizeStates(ctx sdk.Context, queue types.BlockHeightToFinaliz
 	return true
 }
 
+// FastFinalizeRollappStatesUntilStateIndex finalizes states up to the given state index inclusive
+func (k Keeper) FastFinalizeRollappStatesUntilStateIndex(ctx sdk.Context, rollappID string, stateIndex uint64) error {
+	queues, err := k.GetFinalizationQueueByRollapp(ctx, rollappID)
+	if err != nil {
+		return errorsmod.Wrap(err, "get finalization queue")
+	}
+
+	for _, q := range queues {
+		for j, stateInfoIdx := range q.FinalizationQueue {
+			if stateInfoIdx.Index > stateIndex {
+				// Since the queue of queues is sorted, we can break early
+				return nil
+			}
+			err := k.finalizePending(ctx, stateInfoIdx)
+			if err != nil {
+				return errorsmod.Wrap(err, "finalize pending")
+			}
+			q.FinalizationQueue = slices.Delete(q.FinalizationQueue, 0, j)
```

Comment:
are you sure modifying the object we're iterating is not going to mess up with the iteration as now the index which j points to is not the next element (as the underlying finalization queue is changed)?

Reply:
checking!

Reply:
oh man nice catch, dont trust ai! fixed